dubbo-remoting/dubbo-remoting-http12/src/main/java/org/apache/dubbo/remoting/http12/h2/Http2ServerChannelObserver.java

```diff
@@ -74,12 +76,33 @@
 
     @Override
     public void cancel(Throwable throwable) {
+        if (throwable instanceof CancelStreamException) {
+            if (((CancelStreamException) throwable).isCancelByRemote()) {
+                closed = true;
+            }
+        }
+        this.cancellationContext.cancel(throwable);
         long errorCode = 0;
         if (throwable instanceof ErrorCodeHolder) {
             errorCode = ((ErrorCodeHolder) throwable).getErrorCode();
         }
         getHttpChannel().writeResetFrame(errorCode);
-        this.cancellationContext.cancel(throwable);
+    }
+
+    @Override
+    public void onNext(Object data) {
+        if (closed) {
+            return;
+        }
+        super.onNext(data);
```

Comment:
**Optimization Issue**: The onNext method checks if the channel is closed and returns early if it is. This could lead to missed data processing if the close flag is set prematurely. Additionally, there is no handling for the case where data arrives after the channel is marked as closed but before it is actually closed. <br> **Fix**: Implement a more robust mechanism for handling data that arrives after the channel is marked as closed but before it is actually closed. Consider queueing such data for processing or handling it in a way that ensures no data is lost. <br> **Code Suggestion**: 
 ```
 + private final ConcurrentLinkedQueue<Object> pendingData = new ConcurrentLinkedQueue<>();

 + if (closed) {
 +    pendingData.offer(data);
 +    return;
 + }

 + while (!pendingData.isEmpty()) {
 +    Object nextData = pendingData.poll();
 +    super.onNext(nextData);
 + }
 ```